File: bsgroup/bs_group/page/hr_dashboard/hr_dashboard.js
const HRD_METHOD = "bsgroup.bs_group.page.hr_dashboard.hr_dashboard";
const _hrd_state = { tab: "employee" };

frappe.pages["hr-dashboard"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "HR Dashboard",
		single_column: true,
	});

	if (!document.getElementById("hrd-styles")) {
		const style = document.createElement("style");
		style.id = "hrd-styles";
		style.textContent = `
			.hrd-wrap { padding: 20px 24px; background: #f5f7fb; min-height: 100vh; }
			.hrd-header {
				background: linear-gradient(135deg,#111827,#1f2937 55%,#374151);
				color: #fff; border-radius: 20px; padding: 24px 28px; margin-bottom: 18px;
				display: flex; justify-content: space-between; align-items: flex-start;
				flex-wrap: wrap; gap: 14px; box-shadow: 0 10px 25px rgba(15,23,42,0.12);
			}
			.hrd-header h1 { margin: 0; font-size: 26px; font-weight: 800; color: #fff; }
			.hrd-header .sub { font-size: 13px; opacity: .85; margin-top: 4px; }
			.hrd-tabs { display: flex; gap: 8px; flex-wrap: wrap; }
			.hrd-tab {
				background: rgba(255,255,255,0.12); border: 1px solid rgba(255,255,255,0.22);
				color: #fff; padding: 8px 16px; border-radius: 999px; cursor: pointer;
				font-size: 13px; font-weight: 600; white-space: nowrap; transition: background .2s;
			}
			.hrd-tab:hover { background: rgba(255,255,255,0.22); }
			.hrd-tab.active { background: #2563eb; border-color: #2563eb; }
			.hrd-grid4 { display: grid; grid-template-columns: repeat(4,1fr); gap: 16px; margin-bottom: 16px; }
			.hrd-grid2 { display: grid; grid-template-columns: 1.2fr .8fr; gap: 16px; margin-bottom: 16px; }
			@media (max-width: 1100px) { .hrd-grid4 { grid-template-columns: repeat(2,1fr); } .hrd-grid2 { grid-template-columns: 1fr; } }
			@media (max-width: 650px) { .hrd-grid4 { grid-template-columns: 1fr; } }
			.hrd-card { background: #fff; border: 1px solid #e5e7eb; border-radius: 16px; padding: 20px; box-shadow: 0 4px 16px rgba(15,23,42,0.06); }
			.hrd-kpi-label { font-size: 12px; color: #6b7280; text-transform: uppercase; letter-spacing: .04em; }
			.hrd-kpi-value { font-size: 26px; font-weight: 800; margin-top: 8px; color: #1f2937; }
			.hrd-kpi-foot { font-size: 12px; color: #9ca3af; margin-top: 6px; }
			.hrd-sec-title { font-size: 15px; font-weight: 700; margin-bottom: 14px; color: #1f2937; }
			.hrd-status { display: inline-flex; padding: 3px 10px; border-radius: 999px; font-size: 11px; font-weight: 700; }
			.hrd-s-green  { background: #ecfdf3; color: #067647; }
			.hrd-s-amber  { background: #fffaeb; color: #b54708; }
			.hrd-s-red    { background: #fef3f2; color: #b42318; }
			.hrd-s-blue   { background: #eff8ff; color: #175cd3; }
			.hrd-s-gray   { background: #f3f4f6; color: #475467; }
			.hrd-table { width: 100%; border-collapse: collapse; font-size: 13px; }
			.hrd-table th, .hrd-table td { text-align: left; padding: 10px 8px; border-bottom: 1px solid #f3f4f6; }
			.hrd-table th { font-size: 11px; color: #9ca3af; font-weight: 700; text-transform: uppercase; letter-spacing: .04em; }
			.hrd-table tr:hover td { background: #f9fafb; }
			.hrd-empty { color: #9ca3af; text-align: center; padding: 28px 0; font-size: 13px; }
			.hrd-loading { text-align: center; padding: 60px; color: #9ca3af; font-size: 14px; }
			.hrd-link { color: #2563eb; text-decoration: none; font-weight: 600; }
			.hrd-link:hover { text-decoration: underline; }
		`;
		document.head.appendChild(style);
	}

	$(page.body).html(`
		<div class="hrd-wrap">
			<div class="hrd-header">
				<div>
					<h1>HR Dashboard</h1>
					<div class="sub" id="hrd-subtitle">Loading…</div>
				</div>
				<div class="hrd-tabs" id="hrd-tabs"></div>
			</div>
			<div id="hrd-body"><div class="hrd-loading">Loading dashboard…</div></div>
		</div>
	`);

	hrd_boot();
};

function hrd_status_badge(status) {
	const map = { Present: "hrd-s-green", "Half Day": "hrd-s-amber", Absent: "hrd-s-red", "On Leave": "hrd-s-blue", Open: "hrd-s-amber" };
	const cls = map[status] || "hrd-s-gray";
	return `<span class="hrd-status ${cls}">${frappe.utils.escape_html(status || "Not Marked")}</span>`;
}

function hrd_boot() {
	frappe.call({
		method: `${HRD_METHOD}.get_dashboard_context`,
		callback: function (r) {
			const ctx = r.message || {};
			const tabs = [{ key: "employee", label: "My Dashboard" }];
			if (ctx.is_manager) tabs.push({ key: "manager", label: "Manager" });
			if (ctx.is_hr) tabs.push({ key: "hr", label: "HR" });

			$("#hrd-subtitle").text("BS Group · " + (ctx.employee_name || frappe.session.user));
			$("#hrd-tabs").html(
				tabs.map((t) => `<div class="hrd-tab${t.key === _hrd_state.tab ? " active" : ""}" data-tab="${t.key}">${t.label}</div>`).join("")
			);
			$("#hrd-tabs .hrd-tab").on("click", function () {
				_hrd_state.tab = $(this).data("tab");
				$("#hrd-tabs .hrd-tab").removeClass("active");
				$(this).addClass("active");
				hrd_render(_hrd_state.tab);
			});
			if (!tabs.find((t) => t.key === _hrd_state.tab)) _hrd_state.tab = tabs[0].key;
			hrd_render(_hrd_state.tab);
		},
	});
}

function hrd_render(tab) {
	$("#hrd-body").html('<div class="hrd-loading">Fetching live data…</div>');
	if (tab === "employee") hrd_render_employee();
	else if (tab === "manager") hrd_render_manager();
	else if (tab === "hr") hrd_render_hr();
}

function hrd_render_employee() {
	frappe.call({
		method: `${HRD_METHOD}.get_employee_summary`,
		callback: function (r) {
			const d = r.message || {};
			if (!d.has_employee) {
				$("#hrd-body").html('<div class="hrd-card"><div class="hrd-empty">No active Employee record is linked to your user account, so personal attendance/leave data is unavailable. Ask HR to link your User ID on the Employee record.</div></div>');
				return;
			}
			const att = d.attendance_summary || {};
			const leaveRows = (d.leave_balances || []).map((l) => `
				<tr><td>${frappe.utils.escape_html(l.leave_type)}</td><td>${l.entitlement}</td><td>${l.used}</td><td>${l.pending}</td><td><strong>${l.available}</strong></td></tr>
			`).join("") || `<tr><td colspan="5" class="hrd-empty">No leave allocation found</td></tr>`;

			const pendingRows = (d.pending_requests || []).map((p) => `
				<tr><td>${frappe.utils.escape_html(p.leave_type)}</td><td>${frappe.datetime.str_to_user(p.from_date)} – ${frappe.datetime.str_to_user(p.to_date)}</td><td>${hrd_status_badge(p.status)}</td></tr>
			`).join("") || `<tr><td colspan="3" class="hrd-empty">No pending requests</td></tr>`;

			$("#hrd-body").html(`
				<div class="hrd-grid4">
					<div class="hrd-card"><div class="hrd-kpi-label">Today</div><div class="hrd-kpi-value">${d.today_status ? hrd_status_badge(d.today_status === "IN" ? "Present" : d.today_status) : hrd_status_badge("Not Marked")}</div><div class="hrd-kpi-foot">${d.today_time ? frappe.datetime.str_to_user(d.today_time) : ""}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Present (month)</div><div class="hrd-kpi-value">${att.Present || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">On leave (month)</div><div class="hrd-kpi-value">${att["On Leave"] || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Pending requests</div><div class="hrd-kpi-value">${(d.pending_requests || []).length}</div></div>
				</div>
				<div class="hrd-grid2">
					<div class="hrd-card">
						<div class="hrd-sec-title">My Leave</div>
						<table class="hrd-table"><thead><tr><th>Leave Type</th><th>Entitlement</th><th>Used</th><th>Pending</th><th>Available</th></tr></thead><tbody>${leaveRows}</tbody></table>
					</div>
					<div class="hrd-card">
						<div class="hrd-sec-title">Pending Requests</div>
						<table class="hrd-table"><thead><tr><th>Type</th><th>Dates</th><th>Status</th></tr></thead><tbody>${pendingRows}</tbody></table>
					</div>
				</div>
			`);
		},
	});
}

function hrd_render_manager() {
	frappe.call({
		method: `${HRD_METHOD}.get_manager_summary`,
		callback: function (r) {
			const d = r.message || {};
			if (!d.has_reports) {
				$("#hrd-body").html('<div class="hrd-card"><div class="hrd-empty">No direct reports or leave-approver assignments found for your user.</div></div>');
				return;
			}
			const k = d.kpis || {};
			const approvalRows = (d.pending_approvals || []).map((p) => `
				<tr><td>${frappe.utils.escape_html(p.employee_name)}</td><td>${frappe.utils.escape_html(p.leave_type)}</td><td>${frappe.datetime.str_to_user(p.from_date)} – ${frappe.datetime.str_to_user(p.to_date)}</td><td>${hrd_status_badge("Open")}</td><td><a class="hrd-link" href="/app/leave-application/${encodeURIComponent(p.name)}" target="_blank">Review</a></td></tr>
			`).join("") || `<tr><td colspan="5" class="hrd-empty">No pending approvals</td></tr>`;

			const teamRows = (d.team_today || []).map((t) => `
				<tr><td>${frappe.utils.escape_html(t.employee_name)}</td><td>${hrd_status_badge(t.status)}</td></tr>
			`).join("") || `<tr><td colspan="2" class="hrd-empty">No team members</td></tr>`;

			$("#hrd-body").html(`
				<div class="hrd-grid4">
					<div class="hrd-card"><div class="hrd-kpi-label">Present today</div><div class="hrd-kpi-value">${k.present_today || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">On leave</div><div class="hrd-kpi-value">${k.on_leave || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Pending approvals</div><div class="hrd-kpi-value">${k.pending_approvals || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Attendance exceptions</div><div class="hrd-kpi-value">${k.attendance_exceptions || 0}</div></div>
				</div>
				<div class="hrd-grid2">
					<div class="hrd-card">
						<div class="hrd-sec-title">Pending Approvals</div>
						<table class="hrd-table"><thead><tr><th>Employee</th><th>Request</th><th>Dates</th><th>Status</th><th></th></tr></thead><tbody>${approvalRows}</tbody></table>
					</div>
					<div class="hrd-card">
						<div class="hrd-sec-title">Team Today</div>
						<table class="hrd-table"><thead><tr><th>Employee</th><th>Status</th></tr></thead><tbody>${teamRows}</tbody></table>
					</div>
				</div>
			`);
		},
	});
}

function hrd_render_hr() {
	frappe.call({
		method: `${HRD_METHOD}.get_hr_summary`,
		callback: function (r) {
			const d = r.message || {};
			const k = d.kpis || {};
			const recRows = (d.leave_reconciliation || []).map((x) => `
				<tr><td>${frappe.utils.escape_html(x.employee_name)}</td><td>${frappe.utils.escape_html(x.leave_type)}</td><td>${x.ledger_balance}</td><td>${x.pending}</td><td><strong>${x.available}</strong></td><td><span class="hrd-status ${x.available < 0 ? "hrd-s-red" : "hrd-s-amber"}">${frappe.utils.escape_html(x.exception)}</span></td></tr>
			`).join("") || `<tr><td colspan="6" class="hrd-empty">No exceptions found</td></tr>`;

			$("#hrd-body").html(`
				<div class="hrd-grid4">
					<div class="hrd-card"><div class="hrd-kpi-label">Missing leave allocation</div><div class="hrd-kpi-value">${k.missing_leave_allocation || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Missing leave approver</div><div class="hrd-kpi-value">${k.missing_leave_approver || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Active employees</div><div class="hrd-kpi-value">${k.total_active_employees || 0}</div></div>
					<div class="hrd-card"><div class="hrd-kpi-label">Balance exceptions</div><div class="hrd-kpi-value">${(d.leave_reconciliation || []).length}</div></div>
				</div>
				<div class="hrd-card">
					<div class="hrd-sec-title">Leave Balance Reconciliation</div>
					<table class="hrd-table"><thead><tr><th>Employee</th><th>Leave Type</th><th>Ledger Balance</th><th>Pending</th><th>Available</th><th>Exception</th></tr></thead><tbody>${recRows}</tbody></table>
				</div>
			`);
		},
		error: function (e) {
			$("#hrd-body").html(`<div class="hrd-card"><div class="hrd-empty">${e && e.message ? frappe.utils.escape_html(e.message) : "Not permitted."}</div></div>`);
		},
	});
}

File: bsgroup/bs_group/page/hr_dashboard/hr_dashboard.py
import frappe
from frappe.utils import nowdate, get_first_day

HR_ROLES = {"HR User", "HR Manager", "System Manager"}
MANAGER_ROLES = {"HR Manager", "Leave Approver", "System Manager"}


def _get_employee(user):
	return frappe.db.get_value(
		"Employee",
		{"user_id": user, "status": "Active"},
		["name", "employee_name", "department"],
		as_dict=True,
	)


def _is_hr(roles):
	return bool(HR_ROLES & set(roles))


def _get_direct_reports(employee_name, user):
	"""Direct reports via reports_to, plus anyone this user is set as
	Leave Approver for — both are legitimate native ERPNext manager scopes.
	System Manager is an administrative override and sees every active employee."""
	if "System Manager" in frappe.get_roles(user):
		return frappe.get_all(
			"Employee",
			filters={"status": "Active", "name": ["!=", employee_name or ""]},
			fields=["name", "employee_name"],
		)

	reports = {}
	if employee_name:
		for e in frappe.get_all(
			"Employee",
			filters={"reports_to": employee_name, "status": "Active"},
			fields=["name", "employee_name"],
		):
			reports[e.name] = e.employee_name
	for e in frappe.get_all(
		"Employee",
		filters={"leave_approver": user, "status": "Active"},
		fields=["name", "employee_name"],
	):
		reports[e.name] = e.employee_name
	return [{"name": k, "employee_name": v} for k, v in reports.items()]


@frappe.whitelist()
def get_dashboard_context():
	user = frappe.session.user
	roles = frappe.get_roles(user)
	employee = _get_employee(user)
	reports = _get_direct_reports(employee.name if employee else None, user)

	return {
		"employee": employee.name if employee else None,
		"employee_name": employee.employee_name if employee else user,
		"has_employee": bool(employee),
		"is_manager": bool(reports) or bool(MANAGER_ROLES & set(roles)),
		"is_hr": _is_hr(roles),
	}


@frappe.whitelist()
def get_employee_summary():
	employee = _get_employee(frappe.session.user)
	if not employee:
		return {"has_employee": False}

	today = nowdate()
	month_start = get_first_day(today)

	today_attendance = frappe.db.get_value(
		"Employee Checkin",
		{"employee": employee.name, "time": ["between", [f"{today} 00:00:00", f"{today} 23:59:59"]]},
		["log_type", "time"],
		order_by="time asc",
		as_dict=True,
	)

	month_attendance = frappe.get_all(
		"Attendance",
		filters={
			"employee": employee.name,
			"attendance_date": ["between", [month_start, today]],
			"docstatus": ["!=", 2],
		},
		fields=["status"],
	)
	att_counts = {}
	for a in month_attendance:
		att_counts[a.status] = att_counts.get(a.status, 0) + 1

	leave_rows = []
	try:
		from hrms.hr.doctype.leave_application.leave_application import get_leave_details

		details = get_leave_details(employee.name, today)
		for leave_type, d in details.get("leave_allocation", {}).items():
			leave_rows.append({
				"leave_type": leave_type,
				"entitlement": d.get("total_leaves", 0),
				"used": d.get("leaves_taken", 0),
				"pending": d.get("leaves_pending_approval", 0),
				"available": d.get("remaining_leaves", 0),
			})
	except Exception:
		frappe.log_error(title="HR Dashboard: leave balance fetch failed")

	pending_requests = frappe.get_all(
		"Leave Application",
		filters={"employee": employee.name, "status": "Open"},
		fields=["name", "leave_type", "from_date", "to_date", "status"],
		order_by="from_date desc",
		limit_page_length=5,
	)

	return {
		"has_employee": True,
		"employee_name": employee.employee_name,
		"today_status": today_attendance.log_type if today_attendance else None,
		"today_time": today_attendance.time if today_attendance else None,
		"attendance_summary": att_counts,
		"leave_balances": leave_rows,
		"pending_requests": pending_requests,
	}


@frappe.whitelist()
def get_manager_summary():
	user = frappe.session.user
	employee = _get_employee(user)
	reports = _get_direct_reports(employee.name if employee else None, user)
	if not reports:
		return {"has_reports": False}

	report_ids = [r["name"] for r in reports]
	names_by_id = {r["name"]: r["employee_name"] for r in reports}
	today = nowdate()

	pending_leaves = frappe.get_all(
		"Leave Application",
		filters={"employee": ["in", report_ids], "status": "Open"},
		fields=["name", "employee", "leave_type", "from_date", "to_date"],
		order_by="from_date asc",
		limit_page_length=20,
	)
	for r in pending_leaves:
		r["employee_name"] = names_by_id.get(r.employee, r.employee)

	today_attendance = frappe.get_all(
		"Attendance",
		filters={"employee": ["in", report_ids], "attendance_date": today, "docstatus": ["!=", 2]},
		fields=["employee", "status"],
	)
	att_by_emp = {a.employee: a.status for a in today_attendance}

	team_today = []
	present = on_leave = exceptions = 0
	for r in reports:
		status = att_by_emp.get(r["name"], "Not Marked")
		if status == "Present":
			present += 1
		elif status == "On Leave":
			on_leave += 1
		if status == "Not Marked":
			exceptions += 1
		team_today.append({"employee_name": r["employee_name"], "status": status})

	return {
		"has_reports": True,
		"kpis": {
			"present_today": present,
			"on_leave": on_leave,
			"pending_approvals": len(pending_leaves),
			"attendance_exceptions": exceptions,
		},
		"pending_approvals": pending_leaves,
		"team_today": team_today,
	}


@frappe.whitelist()
def get_hr_summary():
	if not _is_hr(frappe.get_roles(frappe.session.user)):
		frappe.throw("Not permitted", frappe.PermissionError)

	# HR is already authorized above; query company-wide regardless of the
	# caller's own User Permission scoping (e.g. a self-scoping Employee
	# permission), which would otherwise silently hide most records from HR.
	active_employees = frappe.get_all(
		"Employee", filters={"status": "Active"}, fields=["name", "employee_name"], ignore_permissions=True
	)
	emp_ids = [e.name for e in active_employees]

	missing_approver = sum(
		1 for e in emp_ids if not frappe.db.get_value("Employee", e, "leave_approver")
	)

	allocated_ids = set(
		frappe.get_all(
			"Leave Allocation",
			filters={"employee": ["in", emp_ids], "docstatus": 1},
			pluck="employee",
			ignore_permissions=True,
		)
	) if emp_ids else set()
	missing_allocation = len([e for e in emp_ids if e not in allocated_ids])

	reconciliation = []
	try:
		from hrms.hr.doctype.leave_application.leave_application import get_leave_details

		today = nowdate()
		for e in active_employees[:50]:
			details = get_leave_details(e.name, today)
			for leave_type, d in details.get("leave_allocation", {}).items():
				pending = d.get("leaves_pending_approval", 0)
				available = d.get("remaining_leaves", 0)
				total = d.get("total_leaves", 0) or 0
				exception = None
				if available < 0:
					exception = f"{abs(available):.1f} days overcommitted"
				elif pending > 0 and total and available <= (0.25 * total):
					exception = "High pending"
				if exception:
					reconciliation.append({
						"employee_name": e.employee_name,
						"leave_type": leave_type,
						"ledger_balance": total - d.get("leaves_taken", 0),
						"pending": pending,
						"available": available,
						"exception": exception,
					})
	except Exception:
		frappe.log_error(title="HR Dashboard: leave reconciliation failed")

	return {
		"kpis": {
			"missing_leave_allocation": missing_allocation,
			"missing_leave_approver": missing_approver,
			"total_active_employees": len(emp_ids),
		},
		"leave_reconciliation": reconciliation[:20],
	}
